src/libs/antares/study/parallel-years.cpp

```diff
@@ -0,0 +1,431 @@
+                        && (p.timeSeriesToRefresh & timeSeriesSolar)
+                        && (y % p.refreshIntervalSolar) == 0);
+    refreshing = refreshing
+                    || ((p.timeSeriesToGenerate & timeSeriesWind)
+                        && (p.timeSeriesToRefresh & timeSeriesWind)
+                        && (y % p.refreshIntervalWind) == 0);
+    refreshing = refreshing
+                    || ((p.timeSeriesToGenerate & timeSeriesHydro)
+                        && (p.timeSeriesToRefresh & timeSeriesHydro)
+                        && (y % p.refreshIntervalHydro) == 0);
+
+    bool haveToRefreshTSThermal 
+        = ((p.timeSeriesToGenerate & timeSeriesThermal)
+                        && (p.timeSeriesToRefresh & timeSeriesThermal)) || thermalTSRefresh_;
+    refreshing
+        = refreshing || (haveToRefreshTSThermal && (y % p.refreshIntervalThermal == 0));
+
+    return refreshing;
+
+}
+
+void SetsOfParallelYearCalculator::buildSetsOfParallelYears()
+{
+    setOfParallelYears* set = nullptr;
+    bool buildNewSet = true;
+    bool foundFirstPerformedYearOfCurrentSet = false;
+    // Gets information on each parallel years set
+    for (uint y = 0; y < p.nbYears; ++y)
+    {
+        unsigned int indexSpace = 999999;
+        bool performCalculations = true;
+        
+        if(p.userPlaylist)
+            performCalculations = p.yearsFilter[y];
+        
+        bool refreshing = isRefreshNeededForCurrentYear(y);
+        buildNewSet = buildNewSet || refreshing;
+
+        if (buildNewSet)
+        {
+            setOfParallelYears setToCreate;
+            setsOfParallelYears.push_back(setToCreate);
+            set = &(setsOfParallelYears.back());
+
+            // Initializations
+            set->nbPerformedYears = 0;
+            set->nbYears = 0;
+            set->regenerateTS = false;
+            set->yearForTSgeneration = 999999;
+
+            // In case we have to regenerate times series before run the current set of parallel
+            // years
+            if (refreshing)
+            {
+                set->regenerateTS = true;
+                set->yearForTSgeneration = y;
+            }
+        }
+
+        set->yearsIndices.push_back(y); 
+        set->nbYears++;
+        set->yearFailed[y] = true;
+        set->isFirstPerformedYearOfASet[y] = false;
+
+        if (performCalculations)
+        {
+            set->setsSizes++;
+            // Another year performed
+            ++nbYearsReallyPerformed;
+
+            // Number of actually performed years in the current set (up to now).
+            set->nbPerformedYears++;
+            // Index of the MC year's space (useful if this year is actually run)
+            indexSpace = set->nbPerformedYears - 1;
+
+            set->isYearPerformed[y] = true;
+            set->performedYearToSpace[y] = indexSpace;
+            set->spaceToPerformedYear[indexSpace] = y;
+
+            if (!foundFirstPerformedYearOfCurrentSet)
```

Comment:
About the snippet : 
```C++
if (!foundFirstPerformedYearOfCurrentSet)
 {
       set->isFirstPerformedYearOfASet[y] = true;
       foundFirstPerformedYearOfCurrentSet = true;
 }

...

// And when we reach the end of the current set, and we have to start another one : 
foundFirstPerformedYearOfCurrentSet = false;

```
This is a complicated mechanism to mark the first performed year of a set.
A simpler way to do this would be to insert few lines above :  
```C++
if (buildNewSet)
 {
       // Start a new set here                
       ...

       // Initializations here
       //...
      if (performCalculations)
           set->isFirstPerformedYearOfASet[y] = true;
     ...
}
```
And that's it. So no need for local variable **foundFirstPerformedYearOfCurrentSet**